packages/astro/src/assets/fonts/vite-plugin-fonts.ts

```diff
@@ -153,163 +75,45 @@ export function fontsPlugin({ settings, sync, logger }: Options): Plugin {
-							},
-							collect,
-						});
-					},
-					root: settings.config.root,
-				});
-				for (const data of fonts) {
-					css += generateFontFace(family.name, data);
-				}
-				const urls = fonts
-					.flatMap((font) => font.src.map((src) => ('originalURL' in src ? src.originalURL : null)))
-					.filter(Boolean);
-
-				const fallbackData = await generateFallbacksCSS({
-					family: family.name,
-					fallbacks: family.fallbacks ?? [],
-					fontURL: urls.at(0) ?? null,
-					getMetricsForFamily: async (name, fontURL) => {
-						let metrics = await fontaine.getMetricsForFamily(name);
-						if (fontURL && !metrics) {
-							// TODO: investigate in using capsize directly (fromBlob) to be able to cache
-							metrics = await fontaine.readMetrics(fontURL);
-						}
-						return metrics;
-					},
-					generateFontFace: fontaine.generateFontFace,
-				});
-
-				if (fallbackData) {
-					css += fallbackData.css;
-					// TODO: generate css var
-				}
-			} else {
-				const { fonts } = await resolveFont(
-					family.name,
-					// We do not merge the defaults, we only provide defaults as a fallback
-					{
-						weights: family.weights ?? DEFAULTS.weights,
-						styles: family.styles ?? DEFAULTS.styles,
-						subsets: family.subsets ?? DEFAULTS.subsets,
-						// No default fallback to be used here
-						fallbacks: family.fallbacks,
-					},
-					// By default, fontaine goes through all providers. We use a different approach
-					// where we specify a provider per font (default to google)
-					[family.provider],
-				);
-
-				for (const data of fonts) {
-					for (const source of data.src) {
-						if ('name' in source) {
-							continue;
-						}
-						source.originalURL = source.url;
-						source.url = proxyURL({
-							value: source.url,
-							// We only use the url for hashing since the service returns urls with a hash already
-							hashString: h64ToString,
-							collect,
-						});
-					}
-					// TODO: support optional as prop
-					css += generateFontFace(family.name, data);
-				}
-
-				const urls = fonts
-					.map((font) => font.src.map((src) => ('originalURL' in src ? src.originalURL : null)))
-					.flat()
-					.filter((url) => typeof url === 'string');
-
-				const fallbackData = await generateFallbacksCSS({
-					family: family.name,
-					fallbacks: family.fallbacks ?? [],
-					fontURL: urls.at(0) ?? null,
-					getMetricsForFamily: async (name, fontURL) => {
-						let metrics = await fontaine.getMetricsForFamily(name);
-						if (fontURL && !metrics) {
-							metrics = await fontaine.readMetrics(fontURL);
-						}
-						return metrics;
-					},
-					generateFontFace: fontaine.generateFontFace,
-				});
+		resolvedMap = new Map();
 
-				if (fallbackData) {
-					css += fallbackData.css;
-					// TODO: generate css var
+		await loadFonts({
+			root: settings.config.root,
+			base: baseUrl,
+			providers: settings.config.experimental.fonts!.providers ?? [],
+			// TS is not smart enough
```

Comment:
Passionate suggestion, having a configuration like this: 

```js
experimental: {
	fonts: {
		providers: []
	}
}
```

It is complicated to maintain, and it could cause problem down the line. I strongly suggest this approach:

```
experimental: {
	fonts: true
},
fonts: {
	providers: []
}
```

Matt switched to this approach with sessions, which made his life easier. I also used this approach for i18n in the past.

In fact, you're kind of forced to use the bang operator here, which is an unsafe code.

Reply:
I need a bang operator because i access `fonts` in a function call but i check for it at the top of the vite plugin. But I see what you mean, I personally prefer it as it is now. Feel free to bring this topic on the RFC!

Reply:
It's just a suggestion. I've seen it before, and I wanted to warn you by talking about my experience. Even the `experimental.svg` brought some difficulties with its types.